brax/physics/colliders.py

```diff
@@ -275,6 +275,125 @@ def apply(cap_a, cap_b, qp_a, qp_b):
 
+class BoxHeightMap:
+  """ A collision between the box corners and a Heihtmap. """
+
+  def __init__(self, config: config_pb2.Config):
+    self.config = config
+    self.pairs = _find_body_pairs(config, 'box', 'heightMap')
+    if not self.pairs:
+      return
+
+    body_idx = {b.name: i for i, b in enumerate(config.bodies)}
+    box_idxs = []
+    heightMap_idxs = []
+    corners = []
+    heights = []
+    sizes = []
+    meshSizes = []
+
+    for box, heightMap in self.pairs:
+      if not heightMap.frozen.all:
+        raise ValueError('active height maps unsupported: %s' % heightMap)
+      for i in range(8):
+        box_idxs.append(body_idx[box.name])
+        heightMap_idxs.append(body_idx[heightMap.name])
+        corner = jnp.array(
+            [i % 2 * 2 - 1, 2 * (i // 4) - 1, i // 2 % 2 * 2 - 1],
+            dtype=jnp.float32)
+        col = box.colliders[0]
+        corner = corner * vec_to_np(col.box.halfsize)
+        corner = math.rotate(corner, euler_to_quat(col.rotation))
+        corner = corner + vec_to_np(col.position)
+        corners.append(corner)
+
+        meshSize = int(jnp.round(jnp.sqrt(len(heightMap.colliders[0].heightMap.data))))
+        if not len(heightMap.colliders[0].heightMap.data) == meshSize ** 2:
+          raise ValueError("The data lenght for an height map should be a perfect square.")
+
+        height = jnp.array(heightMap.colliders[0].heightMap.data).reshape((meshSize,meshSize))
+        heights.append(height)
+
+        sizes.append(heightMap.colliders[0].heightMap.size)
+        meshSizes.append(meshSize)
+
+    body = bodies.Body.from_config(config)
+    self.box = take(body, jnp.array(box_idxs))
+    self.heightMap = take(body, jnp.array(heightMap_idxs))
+    self.corner = jnp.array(corners)
+    self.size = jnp.array(sizes)
+    self.meshSize = jnp.array(meshSizes)
+    self.heights = jnp.array(heights)
+
+
+  def apply(self, qp: QP, dt: float) -> P:
+    """Returns impulse from a collision between box corners and a static height map.
+
+    Note that impulses returned by this function are *not* scaled by dt when
+    applied to parts.  Collision impulses are applied directly as velocity and
+    angular velocity updates.
+
+    Args:
+      qp: Coordinate/velocity frame of the bodies.
+      dt: Integration time step length.
+
+    Returns:
+      dP: Delta velocity to apply to the box bodies in the collision.
+      colliding: Mask for each body: 1 = colliding, 0 = not colliding.
+    """
+    if not self.pairs:
+      return P(jnp.zeros_like(qp.vel), jnp.zeros_like(qp.ang))
+
+    @jax.vmap
+    def apply(box, corner, qp_box, qp_heightMap, size, meshSize, heights):
+      world_pos, vel = math.to_world(qp_box, corner)
+
+      pos = math.inv_rotate(world_pos-qp_heightMap.pos, qp_heightMap.rot)
+
+      uv_pos = (pos[:2])/size*(meshSize-1)
+
+      uv_idx = jnp.floor(uv_pos).astype(jnp.int32) # poorly choosen name for the indexes of the point that form the square around the corner
```

Comment:
nit: is there a better name?     :-)

barring that, just a comment describing this algorithm might allow us to not feel bad about the quality of the name, e.g:

```
# first we find the square in the mesh that enclose the candidate point, with mesh indices ux_idx, ux_udx_u, ...
# then we subdivide the square to find the triangle, with points bla bla
```